routes/dao_coin_exchange_with_fees.go

```diff
@@ -0,0 +1,1602 @@
+		return
+	}
+
+	usdPrice, err := fes.GetQuoteCurrencyPriceInUsd(requestData.QuoteCurrencyPublicKeyBase58Check)
+	if err != nil {
+		_AddBadRequestError(ww, fmt.Sprintf("GetQuoteCurrencyPriceInUsd: Problem getting quote currency price in USD: %v", err))
+		return
+	}
+
+	res := GetQuoteCurrencyPriceInUsdResponse{UsdPrice: usdPrice}
+	if err := json.NewEncoder(ww).Encode(res); err != nil {
+		_AddBadRequestError(ww, fmt.Sprintf("GetQuoteCurrencyPriceInUsd: Problem encoding response: %v", err))
+		return
+	}
+}
+
+func (fes *APIServer) GetQuoteCurrencyPriceInUsd(
+	quoteCurrencyPublicKey string) (string, error) {
+	if IsDesoPkid(quoteCurrencyPublicKey) {
+		desoUsdCents := fes.GetExchangeDeSoPrice()
+		return fmt.Sprintf("%0.9f", float64(desoUsdCents)/100), nil
+	}
+	utxoView, err := fes.backendServer.GetMempool().GetAugmentedUniversalView()
+	if err != nil {
+		return "", fmt.Errorf(
+			"GetQuoteCurrencyPriceInUsd: Error fetching mempool view: %v", err)
+	}
+
+	pkBytes, _, err := lib.Base58CheckDecode(quoteCurrencyPublicKey)
+	if err != nil || len(pkBytes) != btcec.PubKeyBytesLenCompressed {
+		return "", fmt.Errorf(
+			"GetQuoteCurrencyPriceInUsd: Problem decoding public key %s: %v",
+			quoteCurrencyPublicKey, err)
+	}
+
+	existingProfileEntry := utxoView.GetProfileEntryForPublicKey(pkBytes)
+	if existingProfileEntry == nil || existingProfileEntry.IsDeleted() {
+		return "", fmt.Errorf(
+			"GetQuoteCurrencyPriceInUsd: Profile for quote currency public "+
+				"key %v does not exist",
+			quoteCurrencyPublicKey)
+	}
+
+	// If the profile is the dusdc profile then just return 1.0
+	lowerUsername := strings.ToLower(string(existingProfileEntry.Username))
+	if strings.Contains(lowerUsername, "dusdc") {
+		return "1.0", nil
+	} else if strings.Contains(lowerUsername, "focus") ||
+		strings.Contains(lowerUsername, "openfund") {
+
+		desoUsdCents := fes.GetExchangeDeSoPrice()
+		pkid := utxoView.GetPKIDForPublicKey(pkBytes)
+		if pkid == nil {
+			return "", fmt.Errorf("GetQuoteCurrencyPriceInUsd: Error getting pkid for public key %v",
+				quoteCurrencyPublicKey)
+		}
+		ordersBuyingCoin1, err := utxoView.GetAllDAOCoinLimitOrdersForThisDAOCoinPair(
+			&lib.ZeroPKID, pkid.PKID)
+		if err != nil {
+			return "", fmt.Errorf("GetDAOCoinLimitOrders: Error getting limit orders: %v", err)
+		}
+		ordersBuyingCoin2, err := utxoView.GetAllDAOCoinLimitOrdersForThisDAOCoinPair(
+			pkid.PKID, &lib.ZeroPKID)
+		if err != nil {
+			return "", fmt.Errorf("GetDAOCoinLimitOrders: Error getting limit orders: %v", err)
+		}
+		allOrders := append(ordersBuyingCoin1, ordersBuyingCoin2...)
+		// Find the highest bid price and the lowest ask price
+		highestBidPrice := float64(0.0)
+		lowestAskPrice := math.MaxFloat64
+		for _, order := range allOrders {
+			priceStr, err := CalculatePriceStringFromScaledExchangeRate(
+				lib.PkToString(order.BuyingDAOCoinCreatorPKID[:], fes.Params),
+				lib.PkToString(order.SellingDAOCoinCreatorPKID[:], fes.Params),
```

Comment:
I *think* this is ultimately fine, but we really *should* be looking up the public key for PKIDs instead of just converting them.

Reply:
Yeah I left this one as-is because it's an isolated case where it's just checking for deso pkid.